cpp/oneapi/dal/data/array.hpp

```diff
@@ -117,70 +160,132 @@ class array {
     void reset() {
         data_owned_ptr_.reset();
         data_ = std::variant<T*, const T*>();
-        size_ = 0;
+        count_ = 0;
         capacity_ = 0;
     }
 
-    void reset(std::int64_t size) {
-        data_owned_ptr_.reset(new T[size], default_delete{});
-        data_ = data_owned_ptr_.get();
-        size_ = size;
-        capacity_ = size;
+    void reset(std::int64_t count) {
+        reset_impl(detail::host_policy{}, count, detail::host_only_alloc{});
     }
 
     template <typename Deleter>
-    void reset(T* data, std::int64_t size, Deleter&& deleter) {
+    void reset(T* data, std::int64_t count, Deleter&& deleter) {
         // TODO: check input parameters
         data_owned_ptr_.reset(data, std::forward<Deleter>(deleter));
         data_ = data_owned_ptr_.get();
-        size_ = size;
-        capacity_ = size;
+        count_ = count;
+        capacity_ = count;
     }
 
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+    void reset(sycl::queue& queue,
+               std::int64_t count,
+               sycl::usm::alloc kind = sycl::usm::alloc::shared) {
+        reset_impl(queue, count, kind);
+    }
+
+    void reset(sycl::queue& queue,
+               T* data, std::int64_t count,
+               sycl::vector_class<sycl::event> dependencies = {}) {
+        reset(data, count, detail::default_delete<T, decltype(queue)>{ queue });
+        for (auto& event : dependencies) {
+            event.wait();
+        }
+    }
+#endif
+
     template <typename U = T*>
-    void reset_not_owning(U data = nullptr, std::int64_t size = 0) {
+    void reset_not_owning(U data = nullptr, std::int64_t count = 0) {
         data_ = data;
-        size_ = size;
+        count_ = count;
+    }
+
+    void resize(std::int64_t count) {
+        resize_impl(detail::host_policy{}, count, detail::host_only_alloc{});
+    }
+
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+    void resize(sycl::queue& queue,
+                std::int64_t count,
+                sycl::usm::alloc kind = sycl::usm::alloc::shared) {
+        resize_impl(queue, count, kind);
     }
+#endif
 
-    void resize(std::int64_t size) {
+    const T& operator [](std::int64_t index) const {
+        return get_data()[index];
+    }
+
+    T& operator [](std::int64_t index) {
+        return get_mutable_data()[index];
+    }
+
+private:
+    template <typename Policy, typename AllocKind>
```

Comment:
I think `AllocKind` can be replaced by `Allocator`

Reply:
I think `AllocKind` gives more flexibility rather than an `Allocator`:
- Allocator is always an object of a class with a defined set of methods.
- AllocKind is just a flag indicating how we want to allocate memory. Depends on the value of this flag, we can create an appropriate allocator or use a set of malloc/free methods, for example. Without creating any allocator.
I think we should support allocators, but not in this PR. It is not a critical feature. Internally, if we want to have common codebase for Allocators and AllocKinds, we should create allocators for each AllocKind value and use them:
```cpp
template <typename T>
inline T* malloc(const host_seq_policy& policy, std::int64_t count, host_only_alloc) {
   return malloc(policy, count, backend::host_only_allocator{});
}

template <typename T, typename Allocator>
inline T* malloc(const host_policy&, std::int64_t count, Allocator&& alloc) { ... }
```

Reply:
We even can make internal functions like malloc/memset/fill/free accepting only allocators, but for public API we need to have both versions with AllocKind and Allocator.